[AOTInductor] remove CUDA dependency for cpp backend

chenyang78 opened the pull request:
Summary:
Previously, we link against cuda libs even for pure cpp backend.
This caused issues for cases where the inference platform does not
have GPUs. This diff removed cuda dependency for cpp backend.

Reviewed By: bertmaher, muchulee8, mikekgfb

Differential Revision: D49800712




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler

desertfire commented on torch/_inductor/codecache.py:
```diff
@@ -861,6 +857,8 @@ def get_include_and_linking_paths(
         # For those cases, include the lpath and libs command as we do for pytorch above.
         # This approach allows us to only pay for what we use.
         ipaths = cpp_extension.include_paths(cuda) + [sysconfig.get_path("include")]
+        if aot_mode:
```
I don't think we will get to `else` at line 854 for our use case?

chenyang78 replied:
We will get to the ```else``` branch. I hit this with the internal e2e test, where none of clauses from ```include_pytorch```, ```vec_isa != invalid_vec_isa```, ```cuda``` and ```config.cpp.enable_kernel_profile``` is True.

desertfire replied:
`vec_isa == invalid_vec_isa` sounds problematic if we are talking about optimizing cpu models?

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -69,7 +69,7 @@ def load(cls, model, example_inputs, options=None, constraints=None):
                 functions=["run"],
                 extra_ldflags=[so_path],
-                with_cuda=True,  # TODO: change this to not is_cpu
+                with_cuda=False,
```
This should be `with_cuda=not is_cpu,`

chenyang78 replied:
Ah, good catch!

desertfire commented on torch/csrc/inductor/aoti_runtime/model_container.h:
```diff
@@ -299,6 +284,50 @@ class AOTInductorModelContainer {
+          _binary_constants_bin_start + bytes_read,
+          data_size,
+          cudaMemcpyHostToDevice));
+#else // !USE_CUDA
```
Probably don't need the else branch here? My general feeling is `USE_CUDA` and `is_cpu` start to make the code fragmented. Some of the code can be cleaned up a bit, but it seems like using a macro like `USE_CUDA` is inevitable.

chenyang78 replied:
Yeah, I have the same concern. I think ```USE_CUDA``` is inevitable. On the other hand, I think ```is_cpu``` seems to be useful for cases where we had mix-arch execution. I will try to minimize the use of ```is_cpu``` for now and refine it later when we have a better design for supporting multiple arches.

desertfire commented on torch/csrc/inductor/aoti_runtime/cpu_utils.h:
```diff
@@ -0,0 +1,22 @@
+// multi devices.
+#ifndef USE_CUDA
+
+#define AOTI_RUNTIME_DEVICE_CHECK(EXPR)                    \
```
Do we really need this for CPU? If not, I think make it define empty can save some uses of `USE_CUDA` later.

chenyang78 replied:
Previously, I was thinking to have exactly the same empty macro to reduce the uses of ```USE_CUDA```, but changed my mind later. One rationale is that I want to reduce any surprise. For example, I feel turning something like ```AOTI_RUNTIME_DEVICE_CHECK(foo())``` to empty could be a surprise in cpu mode, because function foo may have both cuda and non-cuda side-effect. Instead, having a dedicated macro explicitly for each arch would make the code easier to interpret and reduce the chance of such surprises.

desertfire commented on torch/csrc/inductor/aoti_runtime/device_utils.h:
```diff
@@ -0,0 +1,4 @@
+#pragma once
+
+#include <torch/csrc/inductor/aoti_runtime/cpu_utils.h>
```
The exclusiveness of the sets of macros is not obvious from this file. May be better to just keep this file and merge the contents of the two files into this one.

Commit:
[AOTInductor] remove CUDA dependency for cpp backend (#110409)

Summary:

Previously, we link against cuda libs even for pure cpp backend.
This caused issues for cases where the inference platform does not
have GPUs. This diff removed cuda dependency for cpp backend.

Reviewed By: bertmaher, muchulee8, mikekgfb

Differential Revision: D49800712
